src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -19,22 +23,158 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 	 *
 	 * @return boolean Whether an element was found.
 	 */
-	public function step( $insertion_mode = 'in-body' ) {
-		switch ( $insertion_mode ) {
+	public function step( $insertion_mode = null ) {
+		switch ( $insertion_mode ?: $this->insertion_mode ) {
 			case 'in-body':
 				return $this->step_in_body();
 
 			default:
-				return false;
+				return self::NOT_IMPLEMENTED_YET;
 		}
 	}
 
+	/**
+	 * Parses next element in the 'in head' insertion mode.
+	 *
+	 * Not yet implemented.
+	 *
+	 * @see https://html.spec.whatwg.org/#parsing-main-inhead
+	 *
+	 * @return false
+	 */
+	private function step_in_head() {
+		return self::NOT_IMPLEMENTED_YET;
+	}
+
 	/**
 	 * Parses next element in the 'in body' insertion mode.
 	 *
+	 * @see https://html.spec.whatwg.org/#parsing-main-inbody
+	 *
 	 * @return boolean Whether an element was found.
 	 */
 	private function step_in_body() {
+		ignored:
+		parent::set_bookmark( 'current' );
+		if ( ! $this->next_tag( self::$query ) ) {
+			return false;
+		}
+
+		$tag_name = $this->get_tag();
+		$tag_type = $this->is_tag_closer() ? 'closer' : 'opener';
+
+		/*
+		 * > A start tag whose tag name is "html"
+		 */
+		if ( 'HTML' === $tag_name && 'opener' === $tag_type ) {
+			goto ignored;
+		}
+
+		/*
+		 * > A start tag whose tag name is one of: "base", "basefont", "bgsound",
+		 * > "link", "meta", "noframes", "script", "style", "template", "title"
+		 *
+		 * > An end tag whose tag name is "template"
+		 */
+		if (
+			'opener' === $tag_type && (
+				'BASE' === $tag_name ||
+				'BASEFONT' === $tag_name ||
+				'BGSOUND' === $tag_name ||
+				'LINK' === $tag_name ||
+				'META' === $tag_name ||
+				'NOFRAMES' === $tag_name ||
+				'SCRIPT' === $tag_name ||
+				'STYLE' === $tag_name ||
+				'TEMPLATE' === $tag_name ||
+				'TITLE' === $tag_name
+			) ||
+			(
+				'closer' === $tag_type &&
+				'TEMPLATE' === $tag_name
+			) )
+		{
+			parent::seek( 'current' );
+			$this->insertion_mode = 'in-head';
+			return $this->step();
+		}
+
+		/*
+		 * > A start tag whose tag name is "body"
+		 */
+		if ( 'opener' === $tag_type && 'BODY' === $tag_name ) {
+			goto ignored;
+		}
+
+		/*
+		 * > A start tag whose tag name is "frameset"
+		 */
+		if ( 'opener' === $tag_type && 'FRAMESET' === $tag_name ) {
+			return self::NOT_IMPLEMENTED_YET;
+		}
+
+		/*
+		 * > An end-of-file token
+		 *
+		 * Stop parsing.
+		 */
+
+		/*
+		 * > An end tag whose tag name is "body"
+		 * > An end tag whose tag name is "html"
+		 */
+		if ( 'closer' === $tag_type && ( 'BODY' === $tag_name || 'HTML' === $tag_name ) ) {
+			/*
+			 * > If the stack of open elements does not have a body element in scope, this is a parse error; ignore the token.
+			 *
+			 * @TODO: We didn't construct an open HTML or BODY tag, but we have to make a choice here based on that.
+			 *        Probably need to create these _or_ assume this will always transfer to "after body".
```

Comment:
I'd start with `body` on the stack of open elements and skip it from any stringification – kinda as if we were doing the fragment parsing algorithm with `body` as a context

Reply:
I saw you do this with HTML and think I'll adopt that